https://reviews.llvm.org/D124866: clang-15 does not require hip code to define __noinline__
Just drop the definition in hip headers.
Index: hipamd-rocm-5.2.3/include/hip/amd_detail/host_defines.h
===================================================================
--- hipamd-rocm-5.2.3.orig/include/hip/amd_detail/host_defines.h
+++ hipamd-rocm-5.2.3/include/hip/amd_detail/host_defines.h
@@ -47,7 +47,6 @@ THE SOFTWARE.
 #define __constant__ __attribute__((constant))
 #endif // !__CLANG_HIP_RUNTIME_WRAPPER_INCLUDED__
 
-#define __noinline__ __attribute__((noinline))
 #define __forceinline__ inline __attribute__((always_inline))
 
 #if __HIP_NO_IMAGE_SUPPORT
Index: hipamd-rocm-5.2.3/src/hiprtc/cmake/HIPRTC.cmake
===================================================================
--- hipamd-rocm-5.2.3.orig/src/hiprtc/cmake/HIPRTC.cmake
+++ hipamd-rocm-5.2.3/src/hiprtc/cmake/HIPRTC.cmake
@@ -35,7 +35,6 @@ function(get_hiprtc_macros HIPRTC_DEFINE
 #define __constant__ __attribute__((constant))\n\
 #define __shared__ __attribute__((shared))\n\
 #define __align__(x) __attribute__((aligned(x)))\n\
-#define __noinline__ __attribute__((noinline))\n\
 #define __forceinline__ inline __attribute__((always_inline))\n\
 
 #define launch_bounds_impl0(requiredMaxThreadsPerBlock)                                       \\\n\
